src/algorithms/convex/line_convex_checker.rs

```diff
@@ -0,0 +1,719 @@
+    /// results in another subgraph that can be expressed as line intervals.
+    ///
+    /// If `false` is returned, the `intervals` are left unchanged.
+    #[must_use]
+    pub fn try_extend_intervals(&self, intervals: &mut LineIntervals, node: NodeIndex) -> bool {
+        // Backup the old intervals.
+        let old_intervals = intervals.clone();
+
+        let pos = self.get_position(node);
+        let lines = self.get_lines(node);
+
+        for &line in lines {
+            if let Some(interval) = intervals.get_mut(line) {
+                if pos < interval.min {
+                    if self.line_positions_from(pos, line).nth(1) != Some(interval.min) {
+                        *intervals = old_intervals;
+                        return false;
+                    }
+                    interval.min = pos;
+                } else if pos > interval.max {
+                    if self.line_positions_from(interval.max, line).nth(1) != Some(pos) {
+                        *intervals = old_intervals;
+                        return false;
+                    }
+                    interval.max = pos;
+                }
+            } else {
+                intervals
+                    .0
+                    .push((line, LineInterval { min: pos, max: pos }));
+            }
+        }
+
+        true
+    }
+
+    /// Get the lines a node belongs to.
+    #[inline(always)]
+    pub fn get_lines(&self, node: NodeIndex) -> &[LineIndex] {
+        &self.node_to_pos.get(node).line_indices
+    }
+
+    /// Get the position of a node on its lines.
+    #[inline(always)]
+    pub fn get_position(&self, node: NodeIndex) -> Position {
+        self.node_to_pos.get(node).position
+    }
+
+    /// Get the intervals of positions on each line that are occupied by the nodes.
+    ///
+    /// Return `None` if the nodes do not form contiguous intervals on lines
+    pub fn get_intervals(
+        &self,
+        nodes: impl IntoIterator<Item = NodeIndex>,
+    ) -> Option<LineIntervals> {
+        let nodes = nodes.into_iter();
+
+        // A map from line index to the positions of the nodes on that line.
+        // The map is stored as a SmallVec given that the number of lines is
+        // typically less than 8.
+        let mut line_to_pos = self.get_intervals_scratch_space.borrow_mut();
+        line_to_pos.clear();
+
+        for node in nodes {
+            let pos = self.get_position(node);
+            for &l in self.get_lines(node) {
+                small_map_add(&mut line_to_pos, l, pos);
+            }
+        }
+
+        let mut intervals = LineIntervals::default();
+
+        for &(l, (interval, count)) in line_to_pos.iter() {
+            // Make sure all positions are in the interval are present.
+            let pos_iter = self
+                .line_positions_from(interval.min, l)
+                .take_while(|&p| p <= interval.max);
+            if pos_iter.count() != count {
+                return None;
+            }
```

Comment:
We could skip the linear search here, 

```suggestion
            let start = self
                        .find_index(l, internal.min)
                        .expect("start not on line");
            let end = self
                        .find_index(l, interval.max)
                        .unwrap_or(self.lines[l.as_usize()].len());
            if pos_iter.count() != end-start {
                return None;
            }
```